cirq-core/cirq/circuits/circuit_test.py

```diff
@@ -4726,3 +4726,126 @@ def test_factorize_large_circuit():
     assert len(factors) == 5
     for f, d in zip(factors, desired):
         cirq.testing.assert_has_diagram(f, d)
+
+
+def test_zero_target_operations_go_below_diagram():
+    class CustomOperationAnnotation(cirq.Operation):
+        def __init__(self, text: str):
+            self.text = text
+
+        def with_qubits(self, *new_qubits):
+            raise NotImplementedError()
+
+        @property
+        def qubits(self):
+            return ()
+
+        def _circuit_diagram_info_(self, args) -> str:
+            return self.text
+
+    class CustomOperationAnnotationNoInfo(cirq.Operation):
+        def with_qubits(self, *new_qubits):
+            raise NotImplementedError()
+
+        @property
+        def qubits(self):
+            return ()
+
+        def __str__(self):
+            return "custom!"
+
+    class CustomGateAnnotation(cirq.Gate):
+        def __init__(self, text: str):
+            self.text = text
+
+        def _num_qubits_(self):
+            return 0
+
+        def _circuit_diagram_info_(self, args) -> str:
+            return self.text
+
+    cirq.testing.assert_has_diagram(
+        cirq.Circuit(
+            cirq.Moment(
+                CustomOperationAnnotation("a"),
+                CustomGateAnnotation("b").on(),
+                CustomOperationAnnotation("c"),
+            ),
+            cirq.Moment(
+                CustomOperationAnnotation("e"),
+                CustomOperationAnnotation("d"),
+            ),
+        ),
+        """
+    a   e
+    b   d
+    c
+    """,
+    )
+
+    cirq.testing.assert_has_diagram(
+        cirq.Circuit(
+            cirq.Moment(
+                cirq.H(cirq.LineQubit(0)),
+                CustomOperationAnnotation("a"),
+                cirq.GlobalPhaseOperation(1j),
+            ),
+        ),
+        """
+0: ─────────────H──────
+
+global phase:   0.5π
```

Comment:
Should the global phase operation align with the Hadamard here, like other no-qubit operations?

Reply:
It does align with the Hadamard.

(Your browser might be doing fixed-width wrong.)

Reply:
Sorry, I should have clarified. I meant something like this:
```
0: ─────H──────

        global phase:   0.5π
```

Reply:
Oh. The "global phase" is actually a label aligned with the qubit ids. It's the number that's in the column. Global phase is a 'special' annotation that we supported in the past and I still have it special cased for backwards compatibility.

Reply:
SGTM.